ops/log.py

```diff
@@ -53,6 +55,21 @@ def setup_root_logging(
         debug: if True, write logs to stderr as well as to juju-log.
         exc_stderr: if True, write uncaught exceptions to stderr as well as to juju-log.
     """
+    logging.captureWarnings(True)
+
+    def custom_warning_formatter(
+        message: typing.Union[str, Warning],
+        category: typing.Type[Warning],
+        filename: str,
+        lineno: int,
+        line: typing.Optional[str] = None,
+    ) -> str:
+        module_name = os.path.basename(filename)
+        warning_msg = f'{message}'.replace('\n', ' ')
```

Comment:
Why do we replace `\n` in the message (but not other characters like `\t`)? Presumably warning messages are typically a single line. If not, it's probably okay that they get logged multi-line too. I'd recommend we just pass the `message` straight through -- that's what the default one does.

Reply:
Ah, I see @tonyandrewmeyer's [comment](https://github.com/canonical/operator/issues/1077#issuecomment-2138493463) about this now. However, it doesn't look like the default implementation does this: https://github.com/python/cpython/blob/bd0d97ce3497da35d17daa329c294f4cb87196ee/Lib/warnings.py#L35

Reply:
I may have caused some confusion here, sorry. I mentioned to @IronCore864 that the warning messages should be single line - I think multi-line unstructured log messages are a bad idea, even though we do already have these in debug-log. That was because the default warning format can be multiline (the docs call this out), which is when the code snippet is included, I think.

If we have our own custom format that doesn't have any newlines outside of the message, then I don't think we need to do this. If someone does `warning.warn("hello\nworld")` then that's on them.